connectors/sources/outlook.py

```diff
@@ -0,0 +1,916 @@
+                    )
+                )
+                self.tasks += 1
+
+        await self.queue.put(END_SIGNAL)
+
+    async def _fetch_contacts(self, account, timezone):
+        async for contact in self.client.get_contacts(account=account):
+            await self.queue.put(
+                (
+                    self.doc_formatter.contact_doc_formatter(
+                        contact=contact,
+                        timezone=timezone,
+                    ),
+                    None,
+                )
+            )
+        await self.queue.put(END_SIGNAL)
+
+    async def _fetch_tasks(self, account, timezone):
+        async for task in self.client.get_tasks(account=account):
+            await self.queue.put(
+                (
+                    self.doc_formatter.task_doc_formatter(task=task, timezone=timezone),
+                    None,
+                )
+            )
+
+            if task.has_attachments:
+                await self.fetchers.put(
+                    partial(
+                        self._fetch_attachments, outlook_object=task, timezone=timezone
+                    )
+                )
+                self.tasks += 1
+
+        await self.queue.put(END_SIGNAL)
+
+    async def _fetch_calendars(self, account, timezone):
+        async for calendar in self.client.get_calendars(account=account):
+            await self._enqueue_calendars(
+                calendar=calendar, child_calendar=calendar, timezone=timezone
+            )
+
+        await self.queue.put(END_SIGNAL)
+
+    async def _fetch_child_calendars(self, account, timezone):
+        async for calendar, child_calendar in self.client.get_child_calendars(
+            account=account
+        ):
+            await self._enqueue_calendars(
+                calendar=calendar, child_calendar=child_calendar, timezone=timezone
+            )
+
+        await self.queue.put(END_SIGNAL)
+
+    async def _enqueue_calendars(self, calendar, child_calendar, timezone):
+        await self.queue.put(
+            (
+                self.doc_formatter.calendar_doc_formatter(
+                    calendar=calendar,
+                    child_calendar=str(child_calendar),
+                    timezone=timezone,
+                ),
+                None,
+            )
+        )
+
+        if calendar.has_attachments:
+            await self.fetchers.put(
+                partial(
+                    self._fetch_attachments, outlook_object=calendar, timezone=timezone
+                )
+            )
+            self.tasks += 1
+
+    async def ping(self):
+        """Verify the connection with Outlook"""
+        await self.client.ping()
+        self._logger.info("Successfully connected to Outlook")
+
+    async def _consumer(self):
+        """Async generator to process entries of the queue
+
+        Yields:
+            dictionary: Documents from Outlook.
+        """
+        while self.tasks > 0:
```

Comment:
is it possible that the consumer is faster and has processed all items in the queue, but fetchers are not done yet?

Reply:
To overcome that case, we're first incrementing a task by 1 and once that task is completed then only we're putting an `END SIGNAL` to the queue. 

On the other hand in consumer, we're checking tasks count as well as the END SIGNAL so it won't be possible that the consumer has completed its job first and fetchers are not yet done.